### PR TITLE
8259778: Merge MutableSpace and ImmutableSpace

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change which merges ImmutableSpace into MutableSpace,
eliminating the former.  There were no interesting uses of ImmutableSpace,
other than as the base class for MutableSpace.  The name ImmutableSpace is
kind of a misnomer given that usage.

Testing:
mach5 tier1-3

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259778](https://bugs.openjdk.java.net/browse/JDK-8259778): Merge MutableSpace and ImmutableSpace

### Reviewers
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to e1d95459b05d8ad1326d568c492a2d22ed00bd45
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to e1d95459b05d8ad1326d568c492a2d22ed00bd45
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to e1d95459b05d8ad1326d568c492a2d22ed00bd45

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2271/head:pull/2271`
`$ git checkout pull/2271`

### CODE REVIEW

**Review by sspitsyn — APPROVED** [review]
LGTM

File: src/hotspot/share/gc/parallel/mutableSpace.hpp [hunk 1]
```diff
@@ -25,35 +25,40 @@
...
+// that go with such an assumption.
+//
+// MutableSpace is also responsible for minimizing the
 // page allocation time by having the memory pretouched (with
 // AlwaysPretouch) and for optimizing page placement on NUMA systems
 // by make the underlying region interleaved (with UseNUMA).
 //
-// Invariant: (ImmutableSpace +) bottom() <= top() <= end()
+// Invariant: bottom() <= top() <= end()
 // top() is inclusive and end() is exclusive.
```
Comment (dholmes-ora) [inline src/hotspot/share/gc/parallel/mutableSpace.hpp:47]: If end() is exclusive then shouldn't the invariant be `< end()`?

**Review by dholmes-ora — APPROVED** [review]
This looks good to me.

And I deifnitely agree that a MutableSpace is-not-a ImmutableSpace!

One minor nit below.

Thanks,
David

**Review by tschatzl — APPROVED** [review]
I assume that tier1-3 includes the SA tests :)

Looks good other than that nit.

File: src/hotspot/share/gc/parallel/mutableSpace.hpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/gc/parallel/mutableSpace.hpp:47]: I also think that top() is also exclusive as in other collectors.

@dholmes-ora : e.g. bottom == top == end means the space is empty. These two lines are not disagreeing with each other.

**Comment by tschatzl** [comment]
Hi David,

> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [serviceability-dev](mailto:serviceability-dev@openjdk.java.net):_
> 
> On 28/01/2021 7:09 pm, Thomas Schatzl wrote:
> 
> > On Thu, 28 Jan 2021 05:13:57 GMT, David Holmes <dholmes at openjdk.org> wrote:
> > > > Please review this change which merges ImmutableSpace into MutableSpace,
> > > > eliminating the former.  There were no interesting uses of ImmutableSpace,
> > > > other than as the base class for MutableSpace.  The name ImmutableSpace is
> > > > kind of a misnomer given that usage.
> > > > Testing:
> > > > mach5 tier1-3
> > > 
> > > src/hotspot/share/gc/parallel/mutableSpace.hpp line 47:
> > > > 45: //
> > > > 46: // Invariant: bottom() <= top() <= end()
> > > > 47: // top() is inclusive and end() is exclusive.
> > > 
> > > If end() is exclusive then shouldn't the invariant be `< end()`?
> > 
> > I also think that top() is also exclusive as in other collectors.
> > @dholmes-ora : e.g. bottom == top == end means the space is empty. These two lines are not disagreeing with each other.
> 
> If one is exclusive and one is inclusive then I don't see how they can
> be equal, as that implies they are then both inclusive and exclusive at
> the same time. ?? If end() is exclusive then I would expect an empty
> space to be one where bottom and end are adjacent, not coincident.
> 
> Cheers,
> David

The original comment about top() being inclusive is wrong. top() is also exclusive like in all other collectors as stated elsewhere in my review comment. My "also" in "I also think that top() is also exclusive as in other collectors." probably threw you off after re-reading it, which is wrong. Sorry. 

Maybe some examples help:

bottom = 200, top = 200, end = 200 is an "empty" space (i.e. is of size zero). Whether that empty space is "free" or "fully allocated" or both or neither is another question :)

bottom = 200, top = 200, end = 201 contains one word and is (completely) free (not allocated into at all).

bottom = 200, top = 201, end = 201 contains one word and is full(y allocated).

Top/end are exclusive, and bottom inclusive as does the code assume from what I can tell by quickly looking at it. Still the invariant is bottom <= top <= end in all cases.

Thanks,
  Thomas

**Comment by kimbarrett** [comment]
What @tschatzl said.  The comment saying `top()` is inclusive is simply wrong.  I'll fix it before integrating.

**Comment by kimbarrett** [comment]
Thanks @tschatzl , @sspitsyn , and @dholmes-ora for reviews.

**Pull request closed without merging**